jina/types/arrays/document.py

```diff
@@ -315,20 +329,30 @@ def match(
     ) -> None:
         """Compute embedding based nearest neighbour in `another` for each Document in `self`,
         and store results in `matches`.
+
+        Note:
+            - If metric is 'cosine' it uses the cosine **similarity**.
+            - IF metric is 'euclidean' it uses the euclidean **distance**.
+
         :param darray: the other DocumentArray to match against
         :param metric: the distance metric
         :param limit: the maximum number of matches, when not given
                       all Documents in `another` are considered as matches
         """
         X = np.stack(self.get_attributes('embedding'))
         Y = np.stack(darray.get_attributes('embedding'))
+        if limit is None:
+            limit = len(darray)
         dists = self._compute_dist(X, Y, metric)
+        idx, dist = self._get_sorted_top_k(dists, limit)
 
-        idx, dist = self._get_sorted_top_k(dists, int(limit))
-        for _q, _ids, _dists in zip(docs, idx, dist):
+        for _q, _ids, _dists in zip(self, idx, dist):
             for _id, _dist in zip(_ids, _dists):
-                d = Document(self.docs[int(_id)], copy=True)
-                d.scores['metric'] = _dist
+                d = Document(darray[int(_id)], copy=True)
+                if metric == 'cosine':
+                    d.scores[metric] = 1 - _dist
+                else:
+                    d.scores[metric] = _dist
```

Comment:
parametrize to return `distance` or `similarity`. In case of `euclidean`, similarity can be considered with `1/1+dist`

Reply:
good point. Done